usr/sbin/pkcsstats/pkcsstats.c

```diff
@@ -0,0 +1,867 @@
+{
+    return display_stats(user_id, user_name, (struct display_data *)private);
+}
+
+struct summary_data {
+    CK_ULONG num_slots;
+    CK_SLOT_ID *slots;
+    CK_BYTE *summary_data;
+    CK_ULONG summary_size;
+    CK_SLOT_ID slot_id;
+};
+
+static int summary_mech_cb(CK_MECHANISM_TYPE mech, const char *mech_name,
+                           CK_BYTE *mech_data, CK_ULONG mech_size,
+                           CK_ULONG ofs, void *private)
+{
+    struct summary_data *sd = private;
+    counter_t *slot_counter = (counter_t *)mech_data;
+    counter_t *sum_counter;
+    int i;
+
+    UNUSED(mech);
+    UNUSED(mech_name);
+
+    ofs += sd->slot_id * STAT_SLOT_SIZE;
+    if (ofs + (NUM_SUPPORTED_STRENGTHS + 1) * sizeof(counter_t) >
+                                                        sd->summary_size) {
+        warnx("Internal error: mechanism offset larger than summary size");
+        return 1;
+    }
+
+    sum_counter = (counter_t *)(&sd->summary_data[ofs]);
+
+    for (i = 0; i < NUM_SUPPORTED_STRENGTHS + 1 &&
+                i * sizeof(counter_t) < mech_size; i++)
+        sum_counter[i] += slot_counter[i];
+
+    return 0;
+}
+
+static int summary_slot_cb(CK_SLOT_ID slot_id, CK_BYTE *slot_data,
+                           CK_ULONG slot_size, void *private)
+{
+    int rc;
+    struct summary_data *sd = private;
+
+    sd->slot_id = slot_id;
+
+    rc = for_each_mech(summary_mech_cb, sd, slot_data, slot_size, true);
+
+    return rc < 0 ? 0 : rc;
+}
+
+static int display_summary_cb(int user_id, const char *user_name, void *private)
+{
+    struct summary_data *sd = private;
+    int rc = 0, shm_fd = -1;
+    CK_BYTE *shm_data = NULL;
+    CK_ULONG shm_size = 0;
+
+    rc = open_shm(user_id, user_name, sd->num_slots, &shm_fd, &shm_data,
+                  &shm_size);
+    if (rc != 0)
+        return rc;
+
+    rc = for_all_slots(summary_slot_cb, sd, shm_data, shm_size,
+                       sd->num_slots, sd->slots, false, 0);
+
+    close_shm(shm_fd, shm_data, shm_size);
+    return rc;
+
+}
+
+static int display_summary(struct display_data* dd)
+{
+    struct summary_data sd;
+    int rc = 0;
+
+    sd.num_slots = dd->num_slots;
+    sd.slots = dd->slots;
+    sd.summary_size = dd->num_slots * STAT_SLOT_SIZE;
+    sd.summary_data = calloc(sd.summary_size, 1);
+    if (sd.summary_data == NULL) {
+        warnx("Failed to allocate the summary buffer");
+        return 1;
+    }
+
+    rc = for_all_users(display_summary_cb, &sd);
+    if (rc != 0)
+        goto done;
+
+    printf("Summary (all users):\n\n");
+
+    rc = for_all_slots(display_slot_cb, dd, sd.summary_data, sd.summary_size,
+                       dd->num_slots, dd->slots,
+                       dd->slot_id_specified, dd->slot_id);
+
+done:
+    free(sd.summary_data);
+
+    return rc;
+}
+
+int init_ock(void **dll, CK_FUNCTION_LIST_PTR *func_list)
+{
+    void (*sym_ptr)();
+    CK_RV rc;
+
+    *dll = dlopen("libopencryptoki.so", RTLD_NOW);
```

Comment:
What is the benefit of using library loading here?  We could just directly link against libopencryptoki.so and get rid of all the `dlsym` and the like...

Reply:
well, that's the way all our tools do it....

We could certainly change that at some point in time, but I would not want to do this in this PR. Will certainly need changes in the make files to ensure that the library is build before the tool.